### PR TITLE
New validator

### PR DESCRIPTION
**Pull request by DanielElisenberg**
# NEW VALIDATOR

### What's new?
* Added the new validator package
* Added end to end tests from microdata-validator repo
* Fixed Errors revealed by tests and cleaned up column names etc.
* Changed "validate" to "validate_dataset" to match more with "package_dataset" in the common public API

### What's next?
* Add new unit testing with improvement suggestions from POC PR
* Reassess test coverage and add new end to end tests if necessary

### CODE REVIEW

File: microdata_tools/validation/steps/data_reader.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            "Unsupported measure data type",
+            errors=[f"Unsupported measure data type: {measure_data_type}"],
+        )
+    return csv.ConvertOptions(
+        column_types={
+            "unit_id": pyarrow.string(),
+            "value": pyarrow_data_type,
+            "start": pyarrow.date32(),
+            "stop": pyarrow.date32(),
+            "attributes": pyarrow.string(),
```
Comment (pawbu) [inline microdata_tools/validation/steps/data_reader.py:87]: I see in job-executor that this column is not used when producing the result 🤔

File: microdata_tools/validation/steps/data_reader.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    else:
+        raise ValidationError(
+            "Unsupported measure data type",
+            errors=[f"Unsupported measure data type: {measure_data_type}"],
+        )
+    return csv.ConvertOptions(
+        column_types={
+            "unit_id": pyarrow.string(),
+            "value": pyarrow_data_type,
+            "start": pyarrow.date32(),
```
Comment (pawbu) [inline microdata_tools/validation/steps/data_reader.py:85]: I see that job-executor uses `pyarrow.int16()` type for start and stop dates. Is it OK that we use `date32` here?

File: microdata_tools/validation/steps/data_reader.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+def get_csv_convert_options(measure_data_type: str):
+    pyarrow_data_type = None
+    if measure_data_type == "STRING":
+        pyarrow_data_type = pyarrow.string()
+    elif measure_data_type == "LONG":
+        pyarrow_data_type = pyarrow.int64()
+    elif measure_data_type == "DOUBLE":
+        pyarrow_data_type = pyarrow.float64()
+    elif measure_data_type == "DATE":
+        pyarrow_data_type = pyarrow.date32()
```
Comment (pawbu) [inline microdata_tools/validation/steps/data_reader.py:75]: job-executor uses also `"INSTANT" = pyarrow.int64()` for possible measure data type.

**Review by pawbu — COMMENTED** [review]
Lots of good stuff in here 👍

File: microdata_tools/validation/steps/data_reader.py (hunk 2 above)
Comment (DanielElisenberg) [reply microdata_tools/validation/steps/data_reader.py:85]: the reading as date32 is mostly for the date validation (and automatic isodate => epochdays) pyarrow provides, they get converted to int32 at line 113:
```py
epoch_start = compute.cast(table["start"], target_type=pyarrow.int32())
epoch_stop = compute.cast(table["stop"], target_type=pyarrow.int32())
```

I see that exec uses int16 for the epoch_days columns though, that might be an important distinction. Good catch. I will check this out 👍🏻

File: microdata_tools/validation/steps/data_reader.py (hunk 1 above)
Comment (DanielElisenberg) [reply microdata_tools/validation/steps/data_reader.py:87]: Yeah, it is a completely unnecessary column. It was used 4 times in the first version of the ssb datastore. We still have it in the input model if we want to use it for something again, but currently it does nothing. The csv has 5 columns in the input format so we need to supply something to the schema, but in the `santitize_data` function the attribute variable is thrown away 👍🏻 ( So we get rid of it as soon as possible)

File: microdata_tools/validation/steps/data_reader.py (hunk 3 above)
Comment (DanielElisenberg) [reply microdata_tools/validation/steps/data_reader.py:75]: Yes this is a good catch, but I wonder why date was considered an int16 in the start,stop columns and int64 in the measure column 🤔 Let's discuss this

File: microdata_tools/validation/__init__.py [hunk 4]
```diff
@@ -1,26 +1,167 @@
...
+        raise ValidationError(
+            dataset_name,
+            [
+                f'"{dataset_name}" contains invalid characters. '
+                'Please use only uppercase A-Z, numbers 0-9 or "_"',
+            ],
+        )
+
+
+def get_unit_id_type_for_unit_type(unit_id: str) -> Union[str, None]:
```
[the remark this answers is not in the document]
Comment (DanielElisenberg) [reply microdata_tools/validation/__init__.py:34]: I made a mistake when porting the code so these two functions made no sense 😅 The function is part of the public API as job-executor needs to know the unit_id_type for a given unit_type when pseudonymizing 👍🏻 Fixed them now and added a test to make sure they make sense in the way they should

**Review by pawbu — APPROVED** [review]
🚀

**Review by pawbu — APPROVED** [review]
Looks good 👍

**Pull request merged**